### PR TITLE
[Model][Doc] Add tests, examples and documentation for FLUX.1-schnell

### LINKED ISSUE #574: [New Model]: black-forest-labs/FLUX.1-schnell
### The model to consider.

[****](https://huggingface.co/black-forest-labs/FLUX.1-schnell)
https://huggingface.co/spaces/black-forest-labs/FLUX.1-schnell
Image

### The closest model vllm-omni already supports.

No model from black-forest-labs is supported for now.

### What's your difficulty of supporting the model you want?

_No response_

### Use case and motivation

FLUX.1 [schnell] is the fastest variant in Black Forest Labs' original FLUX.1 text-to-image model family, optimized for rapid generation

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by lishunyang12**
## Purpose

Close #574.

### Scope Change

This PR was originally a full implementation of FLUX.1-schnell (transformer, pipeline, registry). During development, the core model code (`FluxPipeline`, `FluxTransformer2DModel`, registry entries in `registry.py`) was merged into main via the FLUX.1-dev PR. Since FLUX.1-schnell shares the exact same `FluxPipeline` architecture — differing only in default inference steps (4 vs 28) and guidance scale (0 vs 3.5) — the implementation code is already fully functional in upstream.

**What was missing**: there was no way for users to discover FLUX.1-schnell is supported, no examples showing how to run it, no tests validating it works, and no documentation listing its acceleration features. This PR fills those gaps.

### What This PR Adds

- **E2E tests** (`tests/e2e/offline_inference/test_flux1_schnell.py`) — 12 unit tests covering pipeline interface compliance, generation, CFG, batching, dtype consistency, weight loading, and a real `FluxTransformer2DModel` forward pass
- **Online serving examples** (`examples/online_serving/flux/`) — server script, curl example, Python client, README with usage for both schnell and dev variants
- **Documentation**:
  - `docs/models/supported_models.md` — add FLUX.1-schnell to GPU and NPU supported models tables
  - `docs/user_guide/diffusion_acceleration.md` — add acceleration support row (Cache-DiT, CFG-Parallel, Tensor-Parallel)
  - `docs/user_guide/diffusion/parallelism_acceleration.md` — add parallelism support row

## Test Plan

```bash
pytest tests/e2e/offline_inference/test_flux1_schnell.py -v
```

## Test Result

All tests pass (mocked dependencies, no GPU required).
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/registry.py [hunk 1]
```diff
@@ -69,6 +69,11 @@
         "pipeline_sd3",
         "StableDiffusion3Pipeline",
     ),
+    "FLUXPipeline": (
+        "flux",
+        "pipeline_flux",
```
[the remark this answers is not in the document]
Comment (SamitHuang) [reply vllm_omni/diffusion/registry.py]: should change the folder `models/Flux`  to `models/flux`

[the remark this answers is not in the document]
Comment (lishunyang12) [reply vllm_omni/diffusion/registry.py]: Noted, will refine the implementation after the torch profiler tasks

**Review by lishunyang12 — COMMENTED** [review]
The refinement will be left toward the end of Jan.

**Comment by lishunyang12** [comment]
> @david6666666 Can put this model in the roadmap of models, thanks.

@david6666666

**Comment by hsliuustc0106** [comment]
This PR adds support for the FLUX.1-schnell diffusion model. The implementation includes a new transformer architecture, pipeline, and comprehensive test coverage.

**Pros:**
- Comprehensive test suite with 465 lines covering both mocked and real transformer tests
- Clean separation of concerns with dedicated modules for transformer, pipeline, and registry
- Good use of mocking to enable fast testing without downloading weights
- Type annotations are present throughout the code

**Cons:**
- Missing error handling for edge cases (e.g., invalid input dimensions, device mismatches)
- No performance benchmarks or memory usage data despite adding a new model
- Test coverage focuses on happy path; missing tests for error conditions
- No documentation on model-specific requirements or limitations
- Unclear how this integrates with existing diffusion models in the registry

File: tests/e2e/offline_inference/test_flux1_schnell.py [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+        input_ids=torch.zeros((1, 50), dtype=torch.long, device=device),
+        attention_mask=torch.ones((1, 50), dtype=torch.long, device=device),
+    )
+    mock_tokenizer.model_max_length = 512
+
+    # Mock T5 Text Encoder
+    mock_text_encoder = MagicMock()
+    mock_text_encoder.dtype = torch.float32
+    # Output of text encoder must be on the same device as inputs
+    mock_text_encoder.return_value.last_hidden_state = torch.randn(1, 50, 2048, device=device)
```
Comment (hsliuustc0106) [inline tests/e2e/offline_inference/test_flux1_schnell.py]: Hardcoded tensor shape (1, 50, 2048) for text encoder output. Should this be configurable or validated against the actual model's expected dimensions?

File: vllm_omni/diffusion/models/flux/flux_transformer.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    def __init__(
+        self,
+        dim: int,
+        num_heads: int,
+        head_dim: int,
+        qk_norm: bool = True,
+        eps: float = 1e-6,
+        out_bias: bool = True,
+    ):
+        super().__init__()
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/flux/flux_transformer.py:99]: Missing input validation for `hidden_states` dimensions. What happens if the input shape doesn't match the expected format?

File: vllm_omni/diffusion/models/flux/pipeline_flux.py [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+        if not accepts_timesteps:
+            raise ValueError(
+                f"The current scheduler class {scheduler.__class__}'s `set_timesteps` does not support custom"
+                f" timestep schedules. Please check whether you are using the correct scheduler."
+            )
+        scheduler.set_timesteps(timesteps=timesteps, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+        num_inference_steps = len(timesteps)
+    elif sigmas is not None:
+        accept_sigmas = "sigmas" in set(inspect.signature(scheduler.set_timesteps).parameters.keys())
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/flux/pipeline_flux.py:118]: No error handling for model loading failures. Should there be a try-except block with a meaningful error message?

File: vllm_omni/diffusion/models/flux/pipeline_flux.py [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+        prompt = [prompt] if isinstance(prompt, str) else prompt
+        batch_size = len(prompt) if prompt_embeds is None else prompt_embeds.shape[0]
+
+        if prompt_embeds is None:
+            # Tokenize input
+            text_inputs = self.tokenizer(
+                prompt,
+                max_length=max_sequence_length,
+                padding="max_length",
+                truncation=True,
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/flux/pipeline_flux.py]: The denoising loop doesn't have any progress tracking or timeout mechanism. For long-running inference, how do users monitor progress?

File: vllm_omni/diffusion/registry.py [hunk 6]
```diff
@@ -69,6 +69,11 @@
...
+        "FLUXPipeline",
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/registry.py]: Adding FLUX to the registry without version information. How will you handle future FLUX model variants (e.g., FLUX.1-dev vs FLUX.1-schnell)?

File: vllm_omni/diffusion/registry.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lishunyang12) [reply vllm_omni/diffusion/registry.py]: Fixed in 3424bea — renamed `models/Flux` to `models/flux` to match the registry `mod_folder="flux"`.

[the remark this answers is not in the document]
Comment (lishunyang12) [reply vllm_omni/diffusion/registry.py]: Done in 3424bea — renamed to `models/flux`.

File: vllm_omni/diffusion/models/flux/pipeline_flux.py [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+        output_type: str | None = "pil",
+        attention_kwargs: dict[str, Any] | None = None,
+        max_sequence_length: int = 512,
+    ) -> DiffusionOutput:
+        prompt = req.prompt if req.prompt is not None else prompt
+        negative_prompt = req.negative_prompt if req.negative_prompt is not None else negative_prompt
+        height = req.height or 1024  # FLUX.1 default resolution
+        width = req.width or 1024
+        num_inference_steps = req.num_inference_steps or num_inference_steps
+        generator = req.generator or generator
```
[the remark this answers is not in the document]
Comment (lishunyang12) [reply vllm_omni/diffusion/models/flux/pipeline_flux.py]: Fixed in 3424bea — now reads `req.guidance_scale` when provided, falling back to the default 3.5 only if not set.

File: tests/e2e/offline_inference/test_flux1_schnell.py (hunk 2 above)
Comment (lishunyang12) [reply tests/e2e/offline_inference/test_flux1_schnell.py]: The (1, 50, 2048) shape matches T5-base encoder output dimensions. Will add a comment in the test to clarify this is model-specific and should be updated if the text encoder changes.

File: vllm_omni/diffusion/models/flux/flux_transformer.py (hunk 3 above)
Comment (lishunyang12) [reply vllm_omni/diffusion/models/flux/flux_transformer.py:99]: The transformer receives `hidden_states` from `prepare_latents()` which always produces the correct shape. Adding explicit dimension validation would add overhead on every forward call without catching real bugs. The existing `check_inputs()` in the pipeline already validates user-facing inputs.

File: vllm_omni/diffusion/models/flux/pipeline_flux.py (hunk 4 above)
Comment (lishunyang12) [reply vllm_omni/diffusion/models/flux/pipeline_flux.py:118]: Model loading is handled by diffusers `from_pretrained()` which already raises clear errors (e.g. `OSError` for missing files, `RuntimeError` for weight mismatches). Wrapping in try-except would just obscure the root cause.

File: vllm_omni/diffusion/models/flux/pipeline_flux.py (hunk 5 above)
Comment (lishunyang12) [reply vllm_omni/diffusion/models/flux/pipeline_flux.py]: The denoising loop follows the same pattern as other diffusion pipelines in the repo (Wan2.2, Qwen-Image, etc.) — none add progress tracking at the pipeline level. Progress is handled by the orchestrator layer (Omni class). Adding tqdm or callbacks here would break the clean pipeline interface.

File: vllm_omni/diffusion/registry.py (hunk 6 above)
Comment (lishunyang12) [reply vllm_omni/diffusion/registry.py]: FLUX.1-dev is already supported upstream in vllm-omni via `FLUXPipeline` — both schnell and dev share the same pipeline/transformer architecture, just different weights. The registry key `FLUXPipeline` covers both. Variant-specific behavior (e.g. default steps, guidance scale) is handled at the pipeline `forward()` level.

**Comment by lishunyang12** [comment]
@SamitHuang PTAL. I changed the scope of this PR.
